### ISSUE TITLE
Module level type narrowing should affect function definitions inside said module

### ISSUE THREAD

**Issue by JGynther** [microsoft/pyright#10547]
**Is your feature request related to a problem? Please describe.**
Using Pyright language server 1.1.401, the following code spawns a diagnostic

```python
import os

# Attempt to get an environment variable
MY_CONFIG_VALUE = os.environ.get('SOME_ENV_VAR')

# Assert
assert MY_CONFIG_VALUE, 'SOME_ENV_VAR must be set'

# or raise
if not MY_CONFIG_VALUE:
    raise

# MY_CONFIG_VALUE recognized as 'str' here

def takes_string(param: str):
    pass

def use_config():
    takes_string(MY_CONFIG_VALUE)  # <-- Error, MY_CONFIG_VALUE is 'str | None'
```
```
Argument of type "str | None" cannot be assigned to parameter "param" of type "str" in function "takes_string"
  Type "str | None" is not assignable to type "str"
    "None" is not assignable to "str" (Pyright reportArgumentType)
```

However, using a class the type is narrowed
```python
# ... above removed for brevity

class Test:
    test = MY_CONFIG_VALUE # test is recognized as 'str'
```

**Describe the solution you’d like**
It should be possible to narrow the type to str (and pyright does narrow it at module level).

**Comment by erictraut** [comment]
It's not safe to assume that a global (module-scoped) variable will retain its value when accessed at some later time. That's because global variables can be set outside of the module scope.

If you mark the variable as `Final`, then it cannot be modified from outside (without triggering a type checker error). In this case, it's safe for pyright to retain the narrowed type. So, if you simply add a `Final` type qualifier to the `MY_CONFIG_VALUE` declaration, you'll get the behavior you're looking for.

```python
MY_CONFIG_VALUE: Final = ...
```
